crates/red_knot_python_semantic/src/types.rs

```diff
@@ -93,6 +100,24 @@ pub(crate) fn global_symbol_ty<'db>(db: &'db dyn Db, file: File, name: &str) ->
     symbol_ty(db, global_scope(db, file), name)
 }
 
+/// Shorthand for `symbol_ty_by_id` that looks up a global symbol from the context of being in that
+/// module.
+pub(crate) fn global_symbol_lookup<'db>(db: &'db dyn Db, file: File, name: &str) -> Type<'db> {
+    let table = symbol_table(db, global_scope(db, file));
+
+    table
+        .symbol_id_by_name(name)
+        .map(|symbol| {
+            symbol_ty_by_id(
+                db,
+                global_scope(db, file),
+                symbol,
+                SymbolTableLookupContext::GlobalWithinModule,
```

Comment:
So I realized that a simpler approach that doesn't require threading this context through would be to just use `symbol_ty_by_id` (which can always use `Never` and never `Unbound` or `Unknown` as the unbound-ty), and then explicitly use the use-def map here to check directly whether the name may be undeclared and may be unbound, and then if so, look up the builtin type and union that with the symbol public type.